### ISSUE TITLE
Input Format for using T5 for the NER task 

### ISSUE THREAD

**Issue by vkaul11** [google-research/text-to-text-transfer-transformer#339]
I am not getting good results on NER task so I wanted to try T5 but I did not find support for NER.
I was wondering how the data should be formatted. 
What I was thinking was along these lines?
Input: "NER Task: Barack Obama was president of USA ."
Target: "Barack Obama [Entity] USA [Entity]"

OR 
Input: "Barack Obama was president of USA ."
Target: "Barack Obama [Entity] USA [Entity]"
How do we ensure [Entity] is not something that is picked from Sentence Piece  meaning of "Entity"?
Someone in a forum also suggested using Sentinel token (instead of using [Entity]) for this task but I was not sure.
Also can we highlight some text in the Input Data using "/*word/*" Is "*" which was used in wsc data the special character "/*" or sentinel token?
 "Processed input: wsc: The stable was very roomy, with four good stalls; a large swinging window opened into the yard , which made /*it/* pleasant and airy."
Can we couple T5 with CRF because BERT models have known to be coupled with CRF for better NER results?

**Comment by craffel** [comment]
Using a task prefix is not necessary but is generally recommended in case you are doing multi-task training. You could just use "ner:" instead of "NER Task:" though, there is no benefit to including the word "task" there. It also should work fine to just use semicolons or even commas instead of "[Entity]" in the targets; the brackets are probably out-of-domain for the model. In WSC we used * to highlight words. I'm not sure why it wouldn't be working, it could be for lots of reasons - maybe your training data is noisy, maybe it is a tiny training dataset and you are not saving checkpoints often enough so it's overfitting, I really can't say. I don't think it should be necessary to try to combine a CRF with T5.

**Comment by vkaul11** [comment]
Thanks Colin, I think it maybe [Entity] does not work very well. That is what I will try. You recommend that I use 
Target: Barack Obama ; USA ;
or 
Target: Barack Obama * USA *
Is it also ok to use sentinel token for this purpose or that is primarily for masking /denoising unsupervised learning task?

Target: Barack Obama<sentinel token1> USA <sentinel token2>

**Comment by craffel** [comment]
Either of those options could work ok, but simply adding semicolons like Barack Obama; USA; etc would look the most like natural text. You could use sentinels but it would be inconvenient to implement.

**Comment by vkaul11** [comment]
Thanks Colin, one task obviously is to get to exact span of the entities and for that ";" separation sounds good. I will check that first.
However, the final task we have is to get the actual wiki entities as well.
Target: Barack Obama [Q76] USA[Q30]
where all these entity IDs come from Wiki https://www.wikidata.org/wiki/Q30
How can I get this target given that the 100,000 or so wiki-entities will not be part of natural language right?
So I will need to make use new tokens myself and do the pre-training with this task for these tokens which won't be in sentence piece?

**Comment by craffel** [comment]
There is no way the model will be able to output those entity IDs unless it sees them many times during training (either pre-training or fine-tuning) since they apparently have no semantic meaning. The sentencepiece model should be able to tokenize them, but they are otherwise meaningless to the model

**Comment by vkaul11** [comment]
Yes so whenever we have such a classification problem with so many categories it is good to put the entities inside square brackets [Q30] or just use Q30 or use sentinel tokens ? I guess it does not matter as they are new tokens which are not present in natural language. Only way to get this output would be to have a very large data with several instances of each entity ID( not just one instance ?). Is there a minimum number of examples per category we need ( like 5 or 10) so that the model starts to emit these new categorical tokens?

**Comment by craffel** [comment]
Either Q30 or [Q30] should be fine. There is no explicit threshold of the number of examples per category that is necessary; that is up to experimentaion.
